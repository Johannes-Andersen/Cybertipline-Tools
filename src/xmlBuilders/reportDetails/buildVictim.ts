import { XMLBuilder, XMLParser } from 'fast-xml-parser';
import type { Victim } from '../../types';
import { buildDeviceId } from '../common/buildDeviceId';
import { buildEstimatedLocation } from '../common/buildEstimatedLocation';
import { buildIpCapture } from '../common/buildIpCapture';
import { buildPerson } from '../common/buildPerson';

const parser = new XMLParser({
  attributeNamePrefix: '@_',
  ignoreAttributes: false,
  allowBooleanAttributes: true,
});

const builder = new XMLBuilder({
  attributeNamePrefix: '@_',
  ignoreAttributes: false,
  suppressBooleanAttributes: false,
});

export const buildVictim = (victim: Victim, keyName = 'victim'): string => {
  const { estimatedLocation, deviceId, ipCaptureEvent, victimPerson } = victim;

  const deviceIds = deviceId?.map(
    (e) => parser.parse(buildDeviceId(e)).deviceId,
  );
  const ipCaptureEvents = ipCaptureEvent?.map(
    (e) => parser.parse(buildIpCapture(e)).ipCaptureEvent,
  );
  const estimatedLocations = estimatedLocation
    ? parser.parse(buildEstimatedLocation(estimatedLocation)).estimatedLocation
    : undefined;
  const person = parser.parse(buildPerson(victimPerson)).person;

  return builder.build({
    [keyName]: {
      victimPerson: person,
      espIdentifier: victim.espIdentifier,
      espService: victim.espService,
      screenName: victim.screenName,
      displayName: victim.displayName,
      profileUrl: victim.profileUrl,
      ipCaptureEvent: ipCaptureEvents,
      deviceId: deviceIds,
      schoolName: victim.schoolName,
      priorCTReports: victim.priorCTReports,
      estimatedLocation: estimatedLocations,
      additionalInfo: victim.additionalInfo,
    },
  });
};
