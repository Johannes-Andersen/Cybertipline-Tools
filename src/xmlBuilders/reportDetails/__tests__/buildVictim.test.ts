import { describe, expect, it } from 'vitest';
import {
  Country,
  DeviceEventName,
  DeviceIdType,
  IPEventName,
  State,
} from '../../../types';
import type { Victim } from '../../../types';
import { buildVictim } from '../buildVictim';

describe('buildVictim', () => {
  it('should build a victim with all fields', () => {
    const victim: Victim = {
      victimPerson: {
        firstName: 'Jane',
        lastName: 'Doe',
        age: 15,
        dateOfBirth: new Date('2009-05-20'),
        phone: [
          {
            number: '555-3333',
            verified: true,
          },
        ],
        email: [
          {
            email: 'jane.doe@email.com',
            verified: true,
          },
        ],
        address: [
          {
            address: '321 School Lane',
            city: 'Springfield',
            state: State.NewHampshire,
            zipCode: '234',
          },
        ],
      },
      espIdentifier: 'VICTIM789012',
      espService: 'Teen Social Network',
      screenName: 'jane_2009',
      displayName: ['Jane D.', 'JD'],
      profileUrl: ['https://teen-social.com/jane_2009'],
      ipCaptureEvent: [
        {
          ipAddress: '192.168.2.100',
          eventName: IPEventName.Login,
          dateTime: new Date('2024-01-15T13:30:00Z'),
          possibleProxy: false,
        },
      ],
      deviceId: [
        {
          idType: DeviceIdType.IMEI,
          idValue: '987654321012345',
          eventName: DeviceEventName.Login,
          dateTime: new Date('2024-01-15T13:30:00Z'),
        },
      ],
      schoolName: 'Springfield High School',
      priorCTReports: [654321],
      estimatedLocation: {
        city: 'Springfield',
        region: 'Illinois',
        countryCode: Country.UnitedStates,
        verified: true,
        timestamp: new Date('2024-01-15T14:00:00Z'),
      },
      additionalInfo: 'Victim has reported multiple incidents of harassment',
    };

    const result = buildVictim(victim);

    expect(result).toMatchSnapshot();
  });

  it('should build a victim with only required fields', () => {
    const victim: Victim = {
      victimPerson: {
        firstName: 'Alex',
        lastName: 'Smith',
        age: 14,
      },
    };

    const result = buildVictim(victim);

    expect(result).toMatchSnapshot();
  });

  it('should build a victim with online presence details', () => {
    const victim: Victim = {
      victimPerson: {
        firstName: 'Chris',
        lastName: 'Johnson',
        age: 16,
      },
      screenName: 'gamer_chris',
      displayName: ['CJ', 'Chris J.'],
      profileUrl: ['https://gaming-platform.com/gamer_chris'],
      ipCaptureEvent: [
        {
          ipAddress: '192.168.3.100',
          eventName: IPEventName.Login,
        },
        {
          ipAddress: '192.168.3.200',
          eventName: IPEventName.Upload,
        },
      ],
      schoolName: 'Central High School',
      additionalInfo: 'Active on multiple gaming platforms',
    };

    const result = buildVictim(victim);

    expect(result).toMatchSnapshot();
  });

  it('should build a victim with location and device details', () => {
    const victim: Victim = {
      victimPerson: {
        firstName: 'Emma',
        lastName: 'Wilson',
        age: 13,
        dateOfBirth: new Date('2011-08-15'),
      },
      deviceId: [
        {
          idType: DeviceIdType.ICCID,
          idValue: '98765432101234567890',
          eventName: DeviceEventName.Registration,
        },
      ],
      estimatedLocation: {
        city: 'Chicago',
        region: 'Illinois',
        countryCode: Country.UnitedStates,
        verified: false,
      },
      schoolName: 'Lincoln Middle School',
      priorCTReports: [112233],
    };

    const result = buildVictim(victim);

    expect(result).toMatchSnapshot();
  });
});
