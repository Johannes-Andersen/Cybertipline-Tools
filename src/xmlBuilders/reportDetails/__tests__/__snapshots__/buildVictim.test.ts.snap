// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildVictim > should build a victim with all fields 1`] = `"<victim><victimPerson><firstName>Jane</firstName><lastName>Doe</lastName><phone verified="true">555-3333</phone><email verified="true">jane.doe@email.com</email><address><address>321 School Lane</address><city>Springfield</city><zipCode>234</zipCode><state>NH</state></address><age>15</age><dateOfBirth>2009-05-20</dateOfBirth></victimPerson><espIdentifier>VICTIM789012</espIdentifier><espService>Teen Social Network</espService><screenName>jane_2009</screenName><displayName>Jane D.</displayName><displayName>JD</displayName><profileUrl>https://teen-social.com/jane_2009</profileUrl><ipCaptureEvent><ipAddress>192.168.2.100</ipAddress><eventName>Login</eventName><dateTime>2024-01-15T13:30:00.000Z</dateTime><possibleProxy>false</possibleProxy></ipCaptureEvent><deviceId><idType>IMEI</idType><idValue>987654321012345</idValue><eventName>Login</eventName><dateTime>2024-01-15T13:30:00.000Z</dateTime></deviceId><schoolName>Springfield High School</schoolName><priorCTReports>654321</priorCTReports><estimatedLocation verified="true" timestamp="2024-01-15T14:00:00.000Z"><city>Springfield</city><region>Illinois</region><countryCode>US</countryCode></estimatedLocation><additionalInfo>Victim has reported multiple incidents of harassment</additionalInfo></victim>"`;

exports[`buildVictim > should build a victim with location and device details 1`] = `"<victim><victimPerson><firstName>Emma</firstName><lastName>Wilson</lastName><age>13</age><dateOfBirth>2011-08-15</dateOfBirth></victimPerson><deviceId><idType>ICCID</idType><idValue>98765432101234567890</idValue><eventName>Registration</eventName></deviceId><schoolName>Lincoln Middle School</schoolName><priorCTReports>112233</priorCTReports><estimatedLocation verified="false"><city>Chicago</city><region>Illinois</region><countryCode>US</countryCode></estimatedLocation></victim>"`;

exports[`buildVictim > should build a victim with online presence details 1`] = `"<victim><victimPerson><firstName>Chris</firstName><lastName>Johnson</lastName><age>16</age></victimPerson><screenName>gamer_chris</screenName><displayName>CJ</displayName><displayName>Chris J.</displayName><profileUrl>https://gaming-platform.com/gamer_chris</profileUrl><ipCaptureEvent><ipAddress>192.168.3.100</ipAddress><eventName>Login</eventName></ipCaptureEvent><ipCaptureEvent><ipAddress>192.168.3.200</ipAddress><eventName>Upload</eventName></ipCaptureEvent><schoolName>Central High School</schoolName><additionalInfo>Active on multiple gaming platforms</additionalInfo></victim>"`;

exports[`buildVictim > should build a victim with only required fields 1`] = `"<victim><victimPerson><firstName>Alex</firstName><lastName>Smith</lastName><age>14</age></victimPerson></victim>"`;
